### PR TITLE
Bagatur/dedup transformer

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/langchain/langchain/document_transformers/dedup.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            elif self.by_metadata:
+                metadata = doc.metadata
+            else:
+                metadata = {}
+            _doc = Document(page_content=page_content, metadata=metadata)
+            yield _HashedDocument.from_document(_doc)
+
+    def transform_documents(
+        self, documents: Sequence[Document], **kwargs: Any
+    ) -> Sequence[Document]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_transformers/dedup.py:30]: Return type should not be `Sequence` it's not concrete. I suspect we have to fix the base abstraction. (Should be a list for return type)

File: libs/langchain/langchain/document_transformers/dedup.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from typing import Any, Iterable, Iterator, Sequence, Union
+
+from langchain.schema import BaseDocumentTransformer, Document
+from langchain.schema.document import _deduplicate_in_order, _HashedDocument
+
+
+class DedupDocumentTransformer(BaseDocumentTransformer):
+    def __init__(
+        self, by_content: bool = True, by_metadata: Union[bool, Sequence[str]] = False
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_transformers/dedup.py:9]:
```suggestion
        self, *, by_content: bool = True, by_metadata: Union[bool, Sequence[str]] = False
```

File: libs/langchain/langchain/document_transformers/dedup.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ) -> None:
+        self.by_content = by_content
+        self.by_metadata = by_metadata
+
+    def _hashed_documents(
+        self, documents: Iterable[Document]
+    ) -> Iterator[_HashedDocument]:
+        for doc in documents:
+            page_content = doc.page_content if self.by_content else ""
+            if isinstance(self.by_metadata, Sequence):
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_transformers/dedup.py:19]: (fyi `str` is a sequence too -- not sure we should do anything here, but can become an issue if we get used to using it)

File: libs/langchain/langchain/document_transformers/dedup.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+from typing import Any, Iterable, Iterator, Sequence, Union
+
+from langchain.schema import BaseDocumentTransformer, Document
+from langchain.schema.document import _deduplicate_in_order, _HashedDocument
+
+
+class DedupDocumentTransformer(BaseDocumentTransformer):
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_transformers/dedup.py:7]: A short doc-string?

Comment (eyurtsev) [inline libs/langchain/langchain/document_transformers/dedup.py:7]: Perhaps add a unit test? chat gpt can probably write it if given the relevant interfaces

**Review by eyurtsev — APPROVED** [review]

File: libs/langchain/langchain/document_transformers/dedup.py (hunk 1 above)
Comment (baskaryan) [reply libs/langchain/langchain/document_transformers/dedup.py:30]: yea, will save for a sep pr

**Pull request closed without merging**